Benchmarks auto repetitions

MarcelKoch opened the pull request:
This PR enables the automatic deduction of repetition number for the benchmarks. 

For small working sets, the benchmark timings may be too sensitive regarding outliers. With this PR the number of repetitions for the benchmark run is estimated such that the whole benchmark takes >=0.5s. This should result in more stable benchmarks for small problems.

If the repetitions are set to `auto` the warm-up step is skipped.

WIP: The PR enables the new behavior only for the **blas, conversion, spmv** benchmarks. 
Todo:
- [x] enable for solver, preconditioner bencharks
- [x] handle large benchmarks (one repetition takes >0.5s) gracefully

yhmtsai commented:
do we need to estimate the timing or just uses warmup iteration + does repetition until reach the 0.5 sec?
does the warmup (estimate) needs to take 0.25 sec to warmup the device?

Slaedr commented:
What's the reasoning for taking total time as the criterion? With statistics in mind, perhaps it is the number of repetitions that's more important. I like total time for other considerations, such as making sure enough samples are collected by a profiler, but that does not seem to be applicable here.

MarcelKoch commented:
@yhmtsai The repetitions are estimated before the real benchmark run. This way, the timings within the rep loop might be moved outside to reduce the overhead in that loop.
The estimate runs are on the same device as the real benchmark run.

MarcelKoch commented:
@Slaedr In my experience, the runtime for small problems (within L1/L2 cache) can vary quite significantly if only a low number of repetitions is used. With this PR I want to get a more stable average runtime for such problems. I should add, that now the number of repetitions is also exported in the json output.

upsj commented:
How about we start off with warmup + a fixed number of iterations, compute the standard deviation of the runtime in the fixed set, and if that exceeds a certain value relative to the average, we increase the number of iterations by some scheme? I'll check what quickbench, GBench and nonius are doing there.

yhmtsai commented:
for blas, use the prepare for the estimate. Some prepare functions are empty. Is it acceptable?
for spmv, use the apply without refresh the memory.
If we would like to refresh the memory in each timing, we can not move the timing out of loop

MarcelKoch commented:
Concerning the pre and post operations, I was also not sure if they are used correctly/necessary. I guess I will look a bit more into that.

Slaedr commented:
@MarcelKoch Ah, perhaps I see your logic. The total time is a good rule-of-thumb for estimating a good number of repetitions. If something takes very little time, we are likely to need many repeats, which your rule will do. If something takes a lot of time, perhaps it needs only a few repeats, which again is factored in your scheme. There might be some corner cases though, where some poorly-implemented algorithm takes a lot of time, but its runtime is still quite variable. I think what @upsj proposed might be more statistically sound; I remember doing that manually for some my studies in the past.

MarcelKoch commented:
@Slaedr Yes exactly, the runtime is just used as a rule of thumb. I picked up 0.5s as a reasonable minimal runtime somewhere, although I'm not sure where exactly. That runtime can be discussed, if it's too high or too low.

@upsj Running these kinds of statistical tests seems a bit overkill to me. At least in my experience, the variation was quite low for these larger runtimes (>=0.5s), assuming that the machine does not use frequency scaling. If that is enabled, benchmarks can be quite unreliable, so I just ignored that case.
Also, starting with a fixed number of iterations might not be a good choice if one iteration already takes long. That case is also on my to-do list.

upsj commented:
So nonius provides really powerful analysis, but also mainly analyzes tiny pieces of code, so the methods they use there (bootstrapping) don't make much sense on such small sample sizes. 
So I would suggest using either the standard deviation or quartile distance as a measure for variability.

We have a certain variety of runtimes (going from fast to slow: overhead benchmarks, BLAS, SpMV, preconditioners, solvers), so I think it might make sense to have something robust that works on all of them. Requiring SpMVs run for 0.5s looks like a lot of overhead to me, especially since GPUs often have much less variability than CPUs.

upsj commented:
So remembering my statistics lectures back in the day, I think if we want to reduce the standard deviation by a factor of 2, we need to run 4x as many benchmarks, so I guess my suggestion would be
1. compute rel_stddev = stddev / average
2. compute scale = rel_stddev / rel_stddev_limit
3. if scale > 1: run a factor of scale * scale - 1 additional iterations
4. if the standard deviation didn't decrease significantly then, stop anyways and report statistics (quantiles, outliers)

tcojean commented:
I definitely agree with most of what is said here. I remember at the very beginning when we were considering using google benchmarks, the obscure algorithm they used and exactly this scheme of `stopping after reaching 0.5s` without extra control was very weird, that creates an apple to oranges comparison in the timings, i.e. you compare qualitatively very different results depending on problem size and other such things. Particularly, a lot of our solvers and other building blocks go very much past that time (i.e. solvers), so all of this is then not very useful.

IMHO, either you use a same amount of runs for matrices of comparable time (like what we do now), or you use an algorithm like what Tobias tried to outline which you apply equally all the time to reach the same timing accuracy all the time. At the risk of solver benchmarks taking a **ridiculous amount of time**.

tcojean commented:
On the cache effects issue for small problems, another probably more accurate approach is to have proper cache warmup and cache flushing strategies (depending on the context) to stabilize the timings, see this excellent paper on the topic. https://homes.sice.indiana.edu/rcwhaley/papers/timing_SPE08.pdf

MarcelKoch commented:
I guess I don't fully understand the purpose of these benchmarks. I interpreted them as a quick check to see if some changes significantly impact the performance in either way. (With this I mean that the purpose of the benchmarks is not to detect performance changes in the range 5%-1%)

For such a broad comparison, I think this approach is reasonable, both the runtime based and statistics based approach.
On the other hand, a more detailed analysis would require much more effort, as @tcojean already mentioned.

MarcelKoch commented:
Concerning @upsj approach: I'm a bit unsure about the specifics of it. First, what would be a good threshold for the relative stddev? I would guess 1, but I've nothing to support that guess. This is also connected to the underlying distribution, where I'm also not sure which one to assume. The normal distribution would be the easiest choice, but again I've nothing to support that assumption.

upsj commented:
I realized I was actually mixing up two things: stddev of the runtime distribution (which is fixed) and stddev of the mean estimator (which decays like sqrt(n) independent of the distribution). You can get the stddev of the mean estimator by bootstrapping, but that may be overkill here. I need to think about this some more.

As an example I ran a small BLAS benchmarks 1000 times and collected runtimes with 4 and 40 threads
![runtime4](https://user-images.githubusercontent.com/1693511/121771607-a5996000-cb70-11eb-962f-fae39ee7d833.png)
![runtime40](https://user-images.githubusercontent.com/1693511/121771611-a8945080-cb70-11eb-88dc-68b00b824d0a.png)

upsj commented:
I think we should be able to catch all our current cases (quick benchmarks, slow benchmarks) by providing a maximum repetition count used only for the repetition estimate. If the benchmark is slow, then we will choose the number of iterations based on the runtime. If it is fast, we will choose it so we have a sufficient number of runs (let's say 100), but still stay way below 0.5s
That is especially interesting if we are testing hundreds to thousands of small problems, where 0.5s per problem becomes significant. Also I would make both 0.5s and 100 configurable via command-line parameter.
At the moment, we don't seem to have benchmarks that require more sophisticated statistical treatment (though our current `timer` setup allows for it), but that may change when moving to MPI. That is probably a topic for another PR.

upsj commented:
label!

MarcelKoch commented:
I've incorporated a couple of suggestions from this thread.

Now there is more control over the adaptive benchmarking with the additional flags:
- `min_repetitions`
- `max_repetitions`
- `min_runtime`

Also, larger (or rather slower) benchmarks are now handled correctly, i.e. only the minimal requested number of repetitions is executed. Therefore, I've also enabled the adaptive behavior for the solver benchmark. 

For the preconditioner, this approach is not possible, since the total runtime is not updated within the repetitions loop. Therefore, I've added a warning if `-repetitions auto` is used in that case and use the default number of repetitions as a fallback.

On the implementation side, the usage is quite similar to google benchmark, i.e. the following code is valid:
```c++
IterationControl ic(timer);
for(auto status: ic.run()){
  timer->tic()
  // run benchmark
  timer->toc()
}
```
Additionally, `status` may be used to check the number of the current iteration, and if it is the last iteration. 

To clarify, the adaptive benchmarking is only optional and *not* enabled by default.

upsj commented on benchmark/utils/general.hpp:
```diff
@@ -434,4 +460,130 @@ gko::remove_complex<ValueType> compute_max_relative_norm2(
+        iterator begin() const { return iterator{info}; }
+
+        // not used, could potentially used in c++17 (with different type)
+        iterator end() const { return iterator{}; }
```
I guess you are talking about [iterators sentinels](https://www.foonathan.net/2020/03/iterator-sentinel/), which are too beautiful a concept not to mention here :)

upsj commented on benchmark/utils/general.hpp:
```diff
@@ -434,4 +460,130 @@ gko::remove_complex<ValueType> compute_max_relative_norm2(
+        {
+            return cur_it >= min_it &&
+                   (cur_it >= max_it || timer->get_total_time() >= max_runtime);
+        }
```
nit:
```suggestion
        }
        
```

upsj reviewed (approved):
LGTM, great job! I really like the Google Benchmark-like setup.

yhmtsai reviewed (changes requested):
LGTM. Could you also add the doxygen documentation on these classes and functions?

MarcelKoch commented:
@yhmtsai Do you think the documentation is sufficient? If not, could you specify which places need more work?

yhmtsai commented on benchmark/utils/general.hpp:
```diff
@@ -434,4 +460,180 @@ gko::remove_complex<ValueType> compute_max_relative_norm2(
+ * - `cur_it`, containing the current iteration number,
+ * and the methods
+ * - `is_finished`, checks if the benchmark is finished,
+ * - `is_last_iteration`, checks if the current iteration is the last one.
+ *
```
nit
```suggestion
 * - `is_last_iteration`, checks if the current iteration is the last one.
```

yhmtsai commented on benchmark/utils/general.hpp:
```diff
@@ -434,4 +460,180 @@ gko::remove_complex<ValueType> compute_max_relative_norm2(
+         * - the minimum number of iteration is reached
+         * - and either:
+         *   - the maximum number of repetitions is reached
+         *   - the total runtime is above the threshold
+         * @return completeness state of the adaptive run
```
I guess doxygen is fine with this, but we usually have a newline here
nit:
```suggestion
         *   - the total runtime is above the threshold
         *
         * @return completeness state of the adaptive run
```

yhmtsai commented on benchmark/utils/general.hpp:
```diff
@@ -434,4 +460,180 @@ gko::remove_complex<ValueType> compute_max_relative_norm2(
+        /**
+         * checks if the benchmark will stop after the current iteration
+         *
+         * Should only be called after the timing of the current iteration is
+         * complete.
```
you can also put it in `@note` , but it depends on you

yhmtsai reviewed (approved):
LGTM. Thanks for adding these documentation. only a little nit in comments.

yhmtsai commented on benchmark/utils/general.hpp:
```diff
@@ -57,6 +57,8 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <rapidjson/ostreamwrapper.h>
 #include <rapidjson/prettywriter.h>
 
+#include "benchmark/utils/timer.hpp"
```
```suggestion


#include "benchmark/utils/timer.hpp"
```

yhmtsai commented on benchmark/utils/general.hpp:
```diff
@@ -58,6 +58,10 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <rapidjson/prettywriter.h>
 
 
+
```
Sorry, maybe I count the number of lines wrongly.
There are two new line between two header section.

tcojean reviewed (changes requested):
LGTM in general. I would like to clear up the overhead issue we had seen previously first.

upsj commented on benchmark/preconditioner/preconditioner.cpp:
```diff
@@ -159,45 +159,47 @@ void run_preconditioner(const char *precond_name,
             std::unique_ptr<gko::LinOp> precond_op;
-            for (auto i = 0u; i < FLAGS_repetitions; ++i) {
+            for (auto _ : ic_gen.run()) {
+                exec->synchronize();
+                generate_timer->tic();
                 precond_op = precond->generate(system_matrix);
+                generate_timer->toc();
             }
-            generate_timer->toc();
```
In Nonius/Catch2, they measure the overhead of the timer in a separate loop, we could do the same thing here?

tcojean replied:
The problem is that is very dynamic, what changes is one synchronization after many kernel launches vs one synchronization per kernel for the GPU timing. I don't think we can hope to accurately estimate that sadly. Maybe we could change the timer to optionally take a number of iterations information from the outside instead of relying on `tic` and `toc` inside the loop?

upsj replied:
Kind of a two-stage approach, only measuring every n-th time?

tcojean replied:
That would be one way to do it I think, maybe there are other ways as well.

Note that technically, that could also happen for some SpMV and BLAS depending on the parameters I guess. For solvers I think we would not see too much a difference because there is already one synchronization with a few byte being sent in every iteration of the solver (the booleans for the stopping condition).

MarcelKoch replied:
I would go for an approach similar to the first version of this PR, i.e. increasing the number of iterations between two timings until the benchmark is finished. That would result in fewer timings, but still more than one.

Also, that would mean that some part in `IterationControl` can issue timings. Now the question is, should there still be timings outside `IterationControl` or not? To clarify, I think the version with timings outside would look like
```c++
timer->tic();
for(auto _: ic.run()){
  op->apply(...);
}
timer->toc();
auto reps = ic.get_repetitions(); // or something similar
```
or alternatively
```c++
for(auto _: ic.run()){
  op->apply(...);
}
auto runtime = ic.get_average_time();
auto reps = ic.get_repetitions();
```
I think the version without outside timings would be easier to handle, but that's just a guess.

MarcelKoch replied:
I've implemented a version that increases the number of iterations between two timings. For very small examples, this reduces the number of timing call quite a lot, e.g. BLAS benchmark with `n=1000` uses ~18000 repetitions but only 24 timing calls.

Disadvantage: Everything within the `for(auto _: ic.run())` loop is timed. In the solver benchmark this leads to additionally taking the timing of `x_clone = clone(x)`. IMO that should not disturb the benchmark too much.

MarcelKoch replied:
Currently, there is no `exec->synchronize()` call before `timer->tic()`. The CPUTimer automatically synchronizes in its `tic_impl`, should the CUDA, HIP timers also do this?

tcojean replied:
It's interesting that the CPU does this `exec->synchronize()` on `tic`. Anyway that case is mostly for cosmetics because the CPU synchronization should do nothing. I think you can do it for the GPU as well it shouldn't hurt, but it shouldn't change anything either since there shouldn't be any kernel non-synchronized already I think?

I think the solvers can be left like the previous version without timing the `clone(x)` if possible. There shouldn't be a difference in terms of timing overhead due to already doing synchronizations inside the solver apply. The `clone` I believe could be quite impactful since that depends on how reactive the CUDA runtime library is, which in some cases can have weird behaviors.

MarcelKoch replied:
I think it is not possible to reduce the number of timings, if the `clone(x)` should be left out of the timings. That said, the timing overhead didn't matter for the solver benchmark anyways. I will try to figure out how to keep both behaviors (timing in each iteration vs variable intervals between timings).

tcojean replied:
Yes certainly, for solvers I don't expect to win anything here, we will need 1 timing per loop iteration, but as said I think that should be fine. The big issue as you say is we need a way to have both behavior coexisting, hopefully with a not too confusing interface/code.

MarcelKoch replied:
I've added a switch for the behavior change: 
- `ic.run(true)` will manage the timing itself, i.e. everything within the loop will be timed
- `ic.run(false)` does not manage the timing, i.e. the timing has to be issued by the user.

The default case `ic.run()` will use `ic.run(true)`. I think the interface/documentation should be clear enough. (However, the implementation became a bit more messy).

yhmtsai commented:
I did not add the sync in the timer before to keep the sync light.
In Gpu timer, they only need to wait the Event does not need to call device sync.
such that you can tic, toc in the code without a lot of sync there if you only prepare the data outside the loop

MarcelKoch commented:
@yhmtsai All benchmarks added the sync before the `tic` anyways, so I guess there will be no big difference. If I hadn't added the sync in the timer, I would have added it in the `IterationControl` on every `tic/toc`, so the end result should be the same.

yhmtsai commented:
sync before tic should be fine because it is not in the timing, but sync in the toc gives the overhead.
and the sync is for checking the memory is ready.

MarcelKoch commented:
I guess leaving the sync in the `tic` and removing it from `toc` should be fine? That was the original behavior of the benchmarks anyway.

tcojean commented:
Yes the synchronization in the `tic` outside the time area is fine, but the `toc` shouldn't synchronize at least until after the `eventRecord` and `eventSynchronize`, otherwise the overhead would be even bigger.

yhmtsai commented:
I think it maybe only need sync in CPUTimer because it is only way to to make sure the work is ready. (Cpu timer can apply on gpu kernel too.) in GPU timer, we should not have the sync there in tic/toc. If needs to wait some memory movement on cpu, it should be called by user not the timer.

tcojean commented on benchmark/utils/timer.hpp:
```diff
@@ -213,7 +217,6 @@ class CpuTimer : public Timer {
 
     double toc_impl() override
     {
-        exec_->synchronize();
```
CPUs need to keep it though. To explain, there are two ways to time. Let's consider a GPU code (i.e. kernels running).

If you use the CPU based timers (this one), from your POV everything is asynchronous, you submit kernels but don't know anything about when they actually start and when they actually end. To help, what we do is:
+ Synchronize before timing, to ensure nothing is running and we have hope to have a responsive GPU (this one is optional for this phase, but it probably helps in some specific cases).
+ Synchronize before doing the `toc`, so that you know all kernels you launched have completed for real, without this synchronization, because kernel ending and CPUs are non blocking, you would instantly time the end.

When doing GPU timing of a GPU, we are using asynchronous events as well, so here everything happens on the GPU, including the timing, and we use the `eventSynchronize` to ensure get the results back to the CPU. Here we have a pipeline of `eventStart -> kernel -> eventStop -> event2Start -> kernel 2 -> event2Stop`. When doing an `eventSynchronize`, we would still potentially be stalling the pipeline in between kernel and kernel 2 which would give a lesser accuracy than if we did `eventStart -> kernel -> kernel2 -> kerneln -> eventStop`.

yhmtsai commented on benchmark/utils/general.hpp:
```diff
@@ -434,4 +461,232 @@ gko::remove_complex<ValueType> compute_max_relative_norm2(
+
+        void tic()
+        {
+            if (manage_timings) timer->tic();
```
```suggestion
            if (manage_timings) {
                timer->tic();
            }
```
also apply to next one from the gko [ref](https://github.com/ginkgo-project/ginkgo/wiki/Contributing-guidelines#control-flow-constructs)

MarcelKoch replied:
Is that part of the .clang-format specification? If not, perhaps it should be added there.

yhmtsai replied:
it is not in current .clang-format. we only mention it in Contribution guideline.
Does clang-format support this after 6?

MarcelKoch replied:
After a bit of digging, it seems like clang-format (up to 13) still does not support this. But there is an PR for this here: https://reviews.llvm.org/D95168
So it seems like that some future version will support this.

MarcelKoch replied:
There is also a workaround using clang-tidy, but that is over kill (https://stackoverflow.com/a/28437960)

upsj commented on benchmark/spmv/spmv.cpp:
```diff
@@ -123,12 +125,10 @@ void apply_spmv(const char *format_name, std::shared_ptr<gko::Executor> exec,
-                exec->synchronize();
-                tuning_timer->tic();
+            IterationControl ic_tuning{tuning_timer};
+            auto x_clone = clone(x);
+            for (auto _ : ic_tuning.run()) {
```
does that matter? SpMVs should only write to `x` after all.

yhmtsai replied:
no for result.
It depends on what we want the memory status for the benchmark.
Should memory always be a new location (only from software point), or the allocation just be there before the operations.

upsj replied:
at least from the caching standpoint that would probably not make a difference, as the `clone` calls `memcpy`, which might, depending on the implementation, already move the data into cache.

MarcelKoch replied:
I also think that this change should be fine, especially if you consider this as a best-case benchmark, i.e. the data is already in the appropriate caches. Considering the worst-case, i.e. at the beginning of each SpMV the data is not cached, is more difficult in general and would probably require more adjustments, especially wrt Tobias' comment.

MarcelKoch commented on benchmark/solver/solver.cpp:
```diff
@@ -472,9 +475,9 @@ void solve_system(const std::string &solver_name,
+            x_clone = clone(x);
 
             exec->synchronize();
             generate_timer->tic();
```
First thing: you can already use `ic.compute_average_timings` to get the output, since `ic` has internally a copy of the shared ptr for the timer (but only for the apply timer). However, that seems a bit awkward. 

From my viewpoint, IC should not handle any timings in this instance, as `manage_timings` is set to `false`. Adding timings functions to IC would weaken the distinction to the managed case. If the non-managed IC run is requested, the user should take care of the timings.

yhmtsai replied:
It makes sense from this point.
My point is iteration control needs the results from the corresponding timer, so it can be wrong usage.
but I also realize that may make the tic/toc from TimerManager different?
maybe we need to add some comment about that need to use the corresponding timer such that the iteration control can really check it?

MarcelKoch replied:
Perhaps I could add a `get_timer` to IC. That should help to use the correct timer. Otherwise, I can't think of a graceful way of adding tic/toc directly to IC.

yhmtsai replied:
Yeah, it is good.

MarcelKoch commented on benchmark/utils/general.hpp:
````diff
@@ -434,4 +461,232 @@ gko::remove_complex<ValueType> compute_max_relative_norm2(
+ * ```
+ * auto timer = get_timer(...);
+ * IterationControl ic(timer);
+ * for(auto status: ic.[warmup_run|run](manage_timings [default is true])){
````
I have not added the parameter there, as the warmup-run uses always a fixed number of repetitions. I will clarify the documentation.

MarcelKoch commented on benchmark/utils/general.hpp:
```diff
@@ -434,4 +461,232 @@ gko::remove_complex<ValueType> compute_max_relative_norm2(
+
+        double get_total_time() const { return timer->get_total_time(); }
+
+        // emulate shared_ptr behavior
+        const TimerManager *operator->() const { return this; }
+        TimerManager *operator->() { return this; }
```
That was used for being lazy :smile: I will remove that, and adjust the rest accordingly.

tcojean commented on benchmark/utils/timer.hpp:
```diff
@@ -260,6 +264,7 @@ class CudaTimer : public Timer {
 protected:
     void tic_impl() override
     {
+        exec_->synchronize();
```
For the `tic` it doesn't matter right? This is not timed because it's before the `eventRecord` and it could even help in ensuring nothing is running on the GPU when we start running things, like if there was a copy previously (for example, `x_clone`).

yhmtsai replied:
the event should also be after the memcpy if it involves gpu.
For me, ensuring nothing is running on the GPU is the stuff before calling timer, but it does not hurt the timing step.

tcojean reviewed (approved):
LGTM!

tcojean commented on benchmark/utils/general.hpp:
```diff
@@ -434,4 +461,247 @@ gko::remove_complex<ValueType> compute_max_relative_norm2(
+                    cur_info->managed_timer.toc();
+                    stopped = true;
+                    next_timing =
+                        static_cast<IndexType>(std::ceil(next_timing * 1.5));
```
Maybe make the 1.5 controllable for extreme cases?

yhmtsai commented on benchmark/utils/general.hpp:
```diff
@@ -434,4 +461,232 @@ gko::remove_complex<ValueType> compute_max_relative_norm2(
+    run_control warmup_run()
+    {
+        status_warmup_.cur_it = 0;
+        status_warmup_.timer->clear();
```
I realize there's no change from another function, so it is fine.

yhmtsai commented on benchmark/utils/general.hpp:
```diff
@@ -434,4 +465,247 @@ gko::remove_complex<ValueType> compute_max_relative_norm2(
+ * - 'warmup' warmup iterations, applies in fixed and adaptive case
+ * - 'min_repetitions' minimal number of repetitions (adaptive case)
+ * - 'max_repetitions' maximal number of repetitions (adaptive case)
+ * - 'min_runtime' minimal total runtime (adaptive case)
```
miss repetition_growth_factor

yhmtsai commented on benchmark/utils/general.hpp:
```diff
@@ -434,4 +465,247 @@ gko::remove_complex<ValueType> compute_max_relative_norm2(
+ * - `warmup_run()`: controls run defined by `warmup` flag
+ * - `run(bool)`: controls run defined by all other flags
+ * - `get_timer()`: access to underlying timer
+ * Both methods return an object that is to be used in a range-based for loop:
```
not update

yhmtsai reviewed (approved):
LGTM. only two comments for missing documentation

yhmtsai commented on benchmark/utils/general.hpp:
```diff
@@ -434,4 +465,250 @@ gko::remove_complex<ValueType> compute_max_relative_norm2(
+ * - 'min_repetitions' minimal number of repetitions (adaptive case)
+ * - 'max_repetitions' maximal number of repetitions (adaptive case)
+ * - 'min_runtime' minimal total runtime (adaptive case)
+ * - 'repetitions_growth_factor' controls the increase between two successive
```
nit
```suggestion
 * - 'repetition_growth_factor' controls the increase between two successive
```
or change the Gflags name

yhmtsai commented:
need rebase before merge. and you should be able to see there are two pipeline running on gitlab after next push
